run/exec.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package run
+
+import (
+	"io"
+	"os"
+	"os/exec"
+	"os/signal"
+	"strings"
+
+	"github.com/mineiros-io/terramate/errors"
+	"github.com/mineiros-io/terramate/stack"
+	"github.com/rs/zerolog/log"
+)
+
+// Exec will execute the given command on the given stack list
+// During the execution of this function the default behavior
+// for signal handling will be changed so we can wait for the child
+// process to exit before exiting Terramate.
+//
+// If continue on error is true this function will continue to execute
+// commands on stacks even in face of failures, returning an error.L with all errors.
+// If continue on error is false it will return as soon as it finds an error,
+// returning a list with a single error inside.
+func Exec(
+	stacks []stack.S,
+	cmd []string,
+	stdin io.Reader,
+	stdout io.Writer,
+	stderr io.Writer,
+	continueOnError bool,
+) error {
+	logger := log.With().
+		Str("action", "run.Exec()").
+		Str("cmd", strings.Join(cmd, " ")).
+		Logger()
+
+	// Should be at least 1 to avoid losing signals
+	// We are using 3 since it is the number of interrupts
+	// that we handle to do a hard kill, which we could receive
+	// before starting to run a command.
+	const signalsBuffer = 3
+
+	signals := make(chan os.Signal, signalsBuffer)
+	signal.Notify(signals, os.Interrupt)
+	defer signal.Reset(os.Interrupt)
+
+	cmds := make(chan *exec.Cmd)
+	defer close(cmds)
+
+	results := startCmdRunner(cmds)
+
+	errs := errors.L()
+
+	for _, stack := range stacks {
+		cmd := exec.Command(cmd[0], cmd[1:]...)
+		cmd.Dir = stack.HostPath()
+		cmd.Env = os.Environ()
+		cmd.Stdin = stdin
+		cmd.Stdout = stdout
+		cmd.Stderr = stderr
+
+		logger := logger.With().
+			Stringer("stack", stack).
+			Logger()
+
+		logger.Info().Msg("Running")
+
+		if err := cmd.Start(); err != nil {
+			errs.Append(errors.E(stack, err, "running %s", cmd))
+			if continueOnError {
+				continue
+			}
+			return errs.AsError()
+		}
+
+		cmds <- cmd
+		interruptions := 0
+		cmdIsRunning := true
+
+		for cmdIsRunning {
+			select {
+			case sig := <-signals:
+				logger.Trace().
```

Comment:
What do you think of logging them as `info` here? Because this is a kinda useful feedback, so users know that terramate is aware of the SIGINT sent.

Reply:
hmm for the kill I added as Info but on this case it seemed that maybe we would generate too much noise...but I was on the fence. Will change to Info for now :+1: